added blocking capabilities

ParticularMiner opened the pull request:
Hi @Bergvca 

I'm glad you like `red_string_grouper`!  So do I.  😃 

1. As requested, here is a branch with all the matrix-blocking capabilities included.
   
   The following call on the `sec__edgar_companies.csv` data file took 4 minutes (down from 25 minutes on my computer with the last version of `string_grouper`!)

   ```python
   matches = match_strings(companies['Company Name'], n_blocks=(1, 200))
   ```

   It's almost hard to believe.

2. Included is also the option to supply an already initialized `StringGrouper` object to the high-level functions `match_strings`, `match_most_similar`, `group_similar_strings` and `compute_pairwise_similarities`.  
   
   This enables a `StringGrouper` object to be reused thus the corpus can persist between calls as mentioned by @justasojourner in Issue #69.  

3. The README.md at this moment contains links to files on my branch in order to display the images.  After merging (or rebasing) this branch, you will need to do a "search and replace" for the following string before uploading to pypi,org:
"https://raw.githubusercontent.com/ParticularMiner/string_grouper/block/"  &#8594;  "https://raw.githubusercontent.com/Bergvca/string_grouper/master/"

ParticularMiner commented:
@Bergvca 

Unit tests now complete.  Code coverage is also 100%.

ParticularMiner commented:
@Bergvca 

Unit tests now complete.  Code coverage is also 100%.

Bergvca commented:
Hi @ParticularMiner ,

What would be the best order to review the different PR's? Is it best to start with this one? Thanks!

ParticularMiner commented:
Hi @Bergvca ,

Yes.  This one.

ParticularMiner commented:
@Bergvca 

Please don't hesitate to ask me any questions if you have any.  

Cheers.

Bergvca commented on string_grouper/string_grouper.py:
```diff
@@ -212,36 +131,135 @@ def __init__(self, master: pd.Series,
+        self._duplicates = duplicates
+
+    @property
+    def master_id(self):
```
Should we just make these attributes public? So instead of having `self._master_id` we get self.master_id. This eliminates the need for having these setter and property functions.

ParticularMiner replied:
@Bergvca 

Sure.  `master_id` and `duplicates_id` do not need setter and getter functions as neither of them is validated on its own.  I agree that we can get rid of them altogether.  Never mind making them public.  Actually it is best they are always input together with `master` and `duplicates` through the `__init__()` or `reset_data()` functions.

Bergvca commented on string_grouper/string_grouper.py:
```diff
@@ -256,34 +274,224 @@ def n_grams(self, string: str) -> List[str]:
+                left_matrix, right_matrix, nnz_rows[slice(*left_partition)])
+        except OverflowError:
+            # Matrices too big!  Try splitting:
+            left_matrix = None
```
left_matrix / right_matrix variables are not getting used

ParticularMiner replied:
@Bergvca 

Indeed, in the Exception handler, left_matrix and right_matrix are no longer being used.  And since this is a recursive function, I didn’t want them residing in memory hogging the memory space while the function continues to recurse which would lead to degradation in performance.  So I set them to None.  

(Every new call to fit_blockwise_auto() recreates it own copy of these variables and they remain in memory until the end of the recursion.)

Bergvca replied:
Ah I understand now. To make it more clear you could also do `del left_matrix`

Bergvca commented on string_grouper/string_grouper.py:
```diff
@@ -711,3 +942,141 @@ def _validate_id_data(master, duplicates, master_id, duplicates_id):
+    if not corpus:
+        corpus = StringGrouper(string_series_1, string_series_2, **kwargs)
+    else:
+        corpus.reset_data(string_series_1, string_series_2)
```
I'm not sure we should the corpus option to the higher level functions. Basically this is an StringGrouper with a pre-fitted tf-idf vectorizer right? I think there are some dangers here (mainly - the new series can contain ngrams that are not known in the vectorizer), that one should know about when using this. Also it implies deeper knowledge of how the StringGrouper object works. The "High level functions" were created to be able to use the string grouper class without much reading into the code. Once you start inserting your own tf-idf vectorizer I think you are fairly knowledgeable on how the class works, and should use the class functions directly. What do you think? 

I guess i'm not seeing a (common) situation where a user is first building a StringGrouper, and then inserting that in one of the High Level functions, but maybe I'm missing something

Bergvca reviewed (commented):
Hi @ParticularMiner , 

Many thanks for all your work again! It looks very good! As its a pretty big PR, with multiple functionalities involved I just started going through it and adding comments. I still need to wrap my head around the `blocking` functionality and test the code itself, but here are some first comments :).

ParticularMiner commented on string_grouper/string_grouper.py:
```diff
@@ -212,36 +131,135 @@ def __init__(self, master: pd.Series,
+        """
+        self._set_data(master, duplicates, master_id, duplicates_id)
+
+    def clear_data(self):
```
@Bergvca 

I understand your question.  Actually, I’m thinking of importing string_grouper into red_string_grouper.   

And red_string_grouper has a wish-list for string_grouper which includes all these additional attribute functions.  In that way red_string_grouper avoids having to subclass `StringGrouper`.

Let me know if you don’t find this appropriate.

Bergvca replied:
I think your above comment somehow got lost - I think i saw it earlier, but i don't see it anymore...

ParticularMiner commented:
@Bergvca 

Thanks for your comments.

I've updated the code by restoring the high-level functions to what they once were.

Also, I've restored the typing specifications `Optional[pd.Series]`.  Sorry, I shouldn't have removed those.  

All getter and setter methods have been removed, except for `master` and `duplicates`, since they have their own validation routines.

Following my suggestions for `StringGrouper` "shadow member functions" corresponding to the high-level functions which will not rebuild the existing corpus, I've also added the following:

```python
sg = StringGrouper(master, duplicates)
sg.match_strings(new_master, new_duplicates, min_similarity=0.7)
sg.match_most_similar(new_master, new_duplicates, min_similarity=0.7)
sg.group_similar_strings(new_master, min_similarity=0.7)
sg.compute_pairwise_similarities(new_master, new_duplicates, min_similarity=0.7)
```

What do you think?

I'm also awaiting your decision regarding `red_string_grouper`'s wish-list for `string_grouper` to decide whether to finally remove the code reorganization in `StringGrouper`'s `__init__()` function or not.

Bergvca commented:
Hi @ParticularMiner,

I still don't see the use case for these "shadow member functions. ". I understand for example that you want to dedupe first on a large dataset, and later use the same StringGrouper with smaller datasets. In that case however wouldn't you just want to do something like:

```
sg = StringGrouper(master, duplicates)
sg.fit()
result = sg.get_groups()
# now later with a new set of dupes:
sg.set_duplicates(new_duplicates)
result = sg.get_groups()
```

If that make's sense? Maybe I'm just missing something?

ParticularMiner commented:
Hi @Bergvca 

In a `red_string_grouper` use case for example, I wouldn't want to execute the second and third lines of your code:
```python
sg = StringGrouper(master, duplicates)
sg.fit()
result = sg.get_groups()
```
because they would take just too much time for large `master` and `duplicates`.  From empirical observation, matching takes far more time than just building the corpus.   Building the corpus alone takes a very short time (seconds) even for large datasets.  

Besides I wouldn't necessarily need all the matches between `master` and `duplicates`.  For example, in red_string_grouper, I do not match completely new series but subsets of the original `master`.  In that way, I can choose which strings get matched and those that do not:

```python
sg = StringGrouper(master)  # this would build the corpus; no matching  so very fast
sg.set_master(subset_1_of_master)
sg.fit()  # matching here is only between strings in this subset: subset_1_of_master
result_1 = sg.get_groups()
sg.set_master(subset_2_of_master)
sg.fit()  # matching here is only between strings in this subset: subset_2_of_master
result_2 = sg.get_groups()
```

You see, in this way I avoid unwanted and time-consuming matching between strings of  `subset_1_of_master` and strings of `subset_2_of_master`.  At the same time, if I had not maintained the corpus across all the calls, the similarity results would have been inconsistent with each other, because they would have been based on different corpora.

As I mentioned in a previous comment (I don't know if you received it), from a predictive modeling point of view, the new series need not even have all its n-grams in the existing corpus.   Sure, it may not look ideal but it is still a predictive model.









> Hi @ParticularMiner,
> 
> I still don't see the use case for these "shadow member functions. ". I understand for example that you want to dedupe first on a large dataset, and later use the same StringGrouper with smaller datasets. In that case however wouldn't you just want to do something like:
> 
> ```
> sg = StringGrouper(master, duplicates)
> sg.fit()
> result = sg.get_groups()
> # now later with a new set of dupes:
> sg.set_duplicates(new_duplicates)
> result = sg.get_groups()
> ```
> 
> If that make's sense? Maybe I'm just missing something?

ParticularMiner commented:
@Bergvca 

So my above code (of 7 lines) would be condensed into the following 3 lines if a shadow member function was available:
```python
sg = StringGrouper(master)  # this would build the corpus; no matching  so very fast
result_1  = sg.match_strings(subset_1_of_master)
result_2  = sg.match_strings(subset_2_of_master)
```

ParticularMiner commented:
@Bergvca 
> Ah I understand now. To make it more clear you could also do del left_matrix

I agree.  I was ahead of you there.  😄  Did that already.

ParticularMiner commented:
@Bergvca 

See [2<sup>nd</sup> paragraph on page 50 of "Deep Learning for Natural Language Processing: Develop Deep Learning Models" By Jason Brownlee](https://books.google.nl/books?id=_pmoDwAAQBAJ&pg=PA50&lpg=PA50&dq=nlp+new+document+has+words+not+in+vocabulary&source=bl&ots=Iy3Gr87mXV&sig=ACfU3U0GUPkaWKPoKMZZ9ADIdR7CTMOteA&hl=en&sa=X&ved=2ahUKEwid5J6m6qTzAhUNMewKHbBAA8MQ6AF6BAgdEAM#v=onepage&q=nlp%20new%20document%20has%20words%20not%20in%20vocabulary&f=false) for an example of a new dataset with some n-grams not in the corpus.

ParticularMiner commented:
@Bergvca 

So one concrete example I came across while working with two users in Issue #64 and which closely follows my suggested 3-line code snippet above, went like this:

Dataset: A large DataFrame with columns 'Address' and 'State' containing data from the USA.
Goal: To match addresses

The first user reasoned that it was pointless to match any two addresses that were in different states.  So he would group his data by state first and perform the matching on each group.  That made sense to me and also seemed efficient.

On the other hand, the second user didn't bother grouping at all and decided to filter his matching results afterwards.  And it seemed OK to me too.  Indeed, it was the natural way to go!

But as I studied both users' results, it turned out that the similarity scores were different between the users for the same pair of addresses.  Not only that, but for the same `min_similarity`  some matches found in the second user's data were missing in the first user's results!

That's how I remembered that the corpora had been different for each user.  The second user had used a single corpus for all his matches, while the first user had used 51 different corpora (corresponding to 51 US states).  Furthermore, it seemed that the smaller corpora had led to smaller similarity scores, some of which fell below the similarity threshold.

So, to mend the discrepancy between the two results, while hoping to also take advantage of the first user's tremendous gains in efficiency, I thought of a way to recode `StringGrouper` to fix the corpus while preserving its user-interface.  Hence also `red_string_grouper`.

Bergvca commented on string_grouper/string_grouper.py:
```diff
@@ -256,25 +358,196 @@ def n_grams(self, string: str) -> List[str]:
-                matches = StringGrouper._symmetrize_matrix(matches)
+            # the list of matches must be symmetric!
+            # (i.e., if A != B and A matches B; then B matches A)
+            matches = StringGrouper._symmetrize_matrix(matches)
```
_fix_diagonal returns a `csr_matrix`, but `_symmetrize_matrix` expects a `lil_matrix`

ParticularMiner commented:
Hi @Bergvca 
> 
> * When we hit the overflow error, we are potentially already running calculations for a while or not? Should we not output a warning notifying the user of the overflow error, so that in the future it can be solved by setting the `n_blocks`?

I've just added the warnings.  I'm still thinking about the rest ...
